torch/_higher_order_ops/while_loop.py

```diff
@@ -0,0 +1,126 @@
+import torch.utils._pytree as pytree
+
+from torch._C import DispatchKey
+from torch._higher_order_ops.utils import (
+    _has_potential_branch_input_alias,
+    _has_potential_branch_input_mutation,
+    _maybe_run_with_interpreter,
+    autograd_not_implemented,
+    UnsupportedAliasMutationException,
+)
+from torch._ops import HigherOrderOperator
+
+from torch._subclasses.fake_tensor import FakeTensorMode
+from torch.fx.experimental.proxy_tensor import (
+    disable_proxy_modes_tracing,
+    make_fx,
+    ProxyTorchDispatchMode,
+    track_tensor_tree,
+)
+
+
+while_loop_op = HigherOrderOperator("while_loop")
+
+
+def while_loop(cond_fn, body_fn, operands):
+    # TODO(yidi): turn on dynamo in eager-mode
+    assert isinstance(operands, (tuple, list))
+    return while_loop_op(cond_fn, body_fn, operands)
+
+
+@while_loop_op.py_impl(DispatchKey.CompositeExplicitAutograd)
+def while_loop_dense(cond_fn, body_fn, operands):
+    init_val = operands
+    while cond_fn(*init_val):
+        out = body_fn(*init_val)
+        assert isinstance(
+            out, (tuple, list)
+        ), "body_fn should return the same type as operands"
+        assert len(out) == len(
+            init_val
+        ), "body_fn should return the same number of elements as operands"
+        init_val = out
```

Comment:
Should probably just be tuple. More constraining makes it easier to e.g. write a lowering for it in Inductor since there are fewer types to think about.